### ISSUE TITLE
7.1.2 _ink_assert (event == HTTP_TUNNEL_EVENT_DONE)

### ISSUE THREAD

**Issue by hankai17** [apache/trafficserver#5630]
```
(gdb) bt
#0  0x00002ab5214b5207 in raise () from /lib64/libc.so.6
#1  0x00002ab5214b6a38 in abort () from /lib64/libc.so.6
#2  0x00002ab51f513429 in ink_abort (message_format=0x2ab51f524ff0 "%s:%d: failed assertion `%s`") at ink_error.cc:99
#3  0x00002ab51f510dc6 in _ink_assert (expression=0x8071c0 "event == HTTP_TUNNEL_EVENT_DONE", file=0x805892 "HttpSM.cc", line=2952) at ink_assert.cc:37
#4  0x00000000005f7e76 in HttpSM::tunnel_handler (this=0x2ab5abedde00, event=105, data=0x2ab5687517a8) at HttpSM.cc:2952
#5  0x00000000005f6f06 in HttpSM::main_handler (this=0x2ab5abedde00, event=105, data=0x2ab5687517a8) at HttpSM.cc:2698
#6  0x000000000051ae2c in Continuation::handleEvent (this=0x2ab5abedde00, event=105, data=0x2ab5687517a8) at /root/file/trafficserver_7_1_2/iocore/eventsystem/I_Continuation.h:153
#7  0x00000000007af43e in write_signal_and_update (event=105, vc=0x2ab568751600) at UnixNetVConnection.cc:175
#8  0x00000000007b318c in UnixNetVConnection::mainEvent (this=0x2ab568751600, event=1, e=0x2c32a60) at UnixNetVConnection.cc:1263
#9  0x000000000051ae2c in Continuation::handleEvent (this=0x2ab568751600, event=1, data=0x2c32a60) at /root/file/trafficserver_7_1_2/iocore/eventsystem/I_Continuation.h:153
#10 0x00000000007a8013 in InactivityCop::check_inactivity (this=0x2bbfdc0, event=2, e=0x2c32a60) at UnixNet.cc:77
#11 0x000000000051ae2c in Continuation::handleEvent (this=0x2bbfdc0, event=2, data=0x2c32a60) at /root/file/trafficserver_7_1_2/iocore/eventsystem/I_Continuation.h:153
#12 0x00000000007d325c in EThread::process_event (this=0x2ab523835010, e=0x2c32a60, calling_code=2) at UnixEThread.cc:140
#13 0x00000000007d368b in EThread::execute_regular (this=0x2ab523835010) at UnixEThread.cc:221
#14 0x00000000007d388e in EThread::execute (this=0x2ab523835010) at UnixEThread.cc:278
#15 0x00000000007d274a in spawn_thread_internal (a=0x2bbae90) at Thread.cc:84
#16 0x00002ab521853dd5 in start_thread () from /lib64/libpthread.so.0
#17 0x00002ab52157cead in clone () from /lib64/libc.so.6
```

```
(gdb) f 8
#8  0x00000000007b318c in UnixNetVConnection::mainEvent (this=0x2ab568751600, event=1, e=0x2c32a60) at UnixNetVConnection.cc:1263
1263    UnixNetVConnection.cc: No such file or directory.
(gdb) p submit_time
$1 = 1560398564227702796
(gdb) p from_accept_thread
$2 = true
(gdb) p this->f.shutdown
$3 = 1
(gdb) f 10
#10 0x00000000007a8013 in InactivityCop::check_inactivity (this=0x2bbfdc0, event=2, e=0x2c32a60) at UnixNet.cc:77
77      UnixNet.cc: No such file or directory.
(gdb) p now
$4 = 1560398625570610371   (submit_time ~ now just 61s)
```

```
$33 = {{
    fileline = 0x805a6a "HttpSM.cc:629",    state_read_client_request_header  VC_EVENT_READ_READY
    event = 100, 
    reentrancy = 2
  }, {
    fileline = 0x806176 "HttpSM.cc:1355",    state_api_callback
    event = 60000, 
    reentrancy = 3
  }, {
    fileline = 0x8061a1 "HttpSM.cc:1395",    state_api_callout
    event = 60000, 
    reentrancy = 3
  }, {
    fileline = 0x809bc5 "HttpSM.cc:7427",    case HttpTransact::SM_ACTION_DNS_LOOKUP   do_hostdb_lookup();
    event = 65535, 
    reentrancy = 3
  }, {
    fileline = 0x8080e1 "HttpSM.cc:4587",    HttpSM::do_cache_lookup_and_read
    event = 16336, 
    reentrancy = 3
  }, {
    fileline = 0x7fd6b8 "HttpCacheSM.cc:118",   HttpCacheSM::state_cache_open_read  CACHE_EVENT_OPEN_READ
    event = 1102, 
    reentrancy = -1
  }, {
    fileline = 0x806ea2 "HttpSM.cc:2607",    HttpSM::state_cache_open_read   CACHE_EVENT_OPEN_READ
    event = 1102, 
    reentrancy = 1
  }, {
    fileline = 0x808e9a "HttpSM.cc:6160",     HttpSM::setup_cache_read_transfer   append block?
    event = 65535, 
    reentrancy = 1
  }, {
    fileline = 0x807680 "HttpSM.cc:3422",     HttpSM::tunnel_handler_cache_read       VC_EVENT_READ_COMPLETE
    event = 102, 
    reentrancy = 0
  }, {
    fileline = 0x807339 "HttpSM.cc:2948",      HttpSM::tunnel_handler   inactivetimeout
    event = 105, 
    reentrancy = 1
  }, {
    fileline = 0x0, 
    event = 0, 
    reentrancy = 0
  } <repeats 54 times>}
```

proxy.config.http.keep_alive_no_activity_timeout_in INT 60
proxy.config.http.transaction_no_activity_timeout_in INT 30

**Comment by hankai17** [comment]
so strange: f.shutdown is 1

**Comment by scw00** [comment]
So Do you have any plugin has TSVConnShutdown function call ？

**Comment by hankai17** [comment]
@scw00 Very thanks.
I used the default background_fetch.so and another one plugin that set cache key.
Only In background_fetch.so, I found that 
```
case TS_EVENT_VCONN_WRITE_COMPLETE:
    // TSVConnShutdown(data->vc, 0, 1);
    // TSVIOReenable(data->w_vio);
    TSDebug(PLUGIN_NAME, "Write Complete");
    break;
```
But "TSVConnShutdown(data->vc, 0, 1)" is commented out. That is also in master branch.
It was seem that the request hited the cache disk. But I can not get more info about it.
Here is the object HttpSM

[httpsm.txt](https://github.com/apache/trafficserver/files/3293015/httpsm.txt)
